alerta/plugins/blackout.py

```diff
@@ -19,13 +17,17 @@ class BlackoutHandler(PluginBase):
     any plugins for further notification.
     """
 
-    def pre_receive(self, alert):
+    def pre_receive(self, alert, **kwargs):
+
+        config = kwargs['config']
 
-        if app.config['ALARM_MODEL'] == 'ALERTA':
+        if config['ALARM_MODEL'] == 'ALERTA':
             status = 'blackout'
         else:
             status = 'OOSRV'  # ISA_18_2
 
+        NOTIFICATION_BLACKOUT = os.environ.get('NOTIFICATION_BLACKOUT') or config.get('NOTIFICATION_BLACKOUT', True)
```

Comment:
Minor:
I was wondering if we can create a function like `get_config` added in `PluginBase` class which can extract configuration from `kwargs['config']`. 
This will ease refactoring of all the existing Alerta plugins. So, adding below code:

```
config = kwargs['config'] 
...
NOTIFICATION_BLACKOUT = os.environ.get('NOTIFICATION_BLACKOUT') or config.get('NOTIFICATION_BLACKOUT', True)
```

will just become

```
NOTIFICATION_BLACKOUT = self.get_config('NOTIFICATION_BLACKOUT')
```

Reply:
Yes, I tried exactly this but couldn't get it work. If you've got any ideas let me know. 😢

Reply:
Hmm I see. We are transferring config as kwargs, so don't have much freedom, what about:

```
    def get_config(self, key, default, **kwargs):
        config = kwargs['config'] 
        return config.get(key) or os.environ.get(key) or default
```

And usage as:

```
    NOTIFICATION_BLACKOUT = self.get_config('NOTIFICATION_BLACKOUT', True, kwargs)
```

Also, I just realised the ordering is wrong in original commit. i.e. kwargs config should have higher preference, followed by env value and then default. (earlier was env > kwargs)

Reply:
Thanks a lot, the `get_config` method looks awesome, even better with `type` parameter. :D

Reply:
Good suggestion. The helper method I've come up with is ...
```
def get_config(self, key, default=None, type=None)
```
And usage is ...
```
self.kwargs = kwargs

NOTIFICATION_BLACKOUT = self.get_config('NOTIFICATION_BLACKOUT', default=True, type=bool)
```
Note: you must initialise `self.kwargs` yourself if you want to use `self.get_config()`. It will raise an exception if you try to use the helper without initialising first.

If you supply a `type` it will convert any environment variable into that type. Supported types are `bool`, `int`, `float`, `list`, JSON-encoded `dict` with a default of `str`.

Reply:
Oh, and precedence order is environment variable first, then any defined config, then default. I'm not sure why you would want to reverse it. If you disagree can you please elaborate?

Reply:
Yes, I think defined config should be first because if someone is especially passing config in `routing.py`, he would especially want it be used & knows what he is doing. While the env variable/default for the fallback alerts. (at least how I am using right now)

Otherwise, the feature will not be useful/workable if whole Alerta have an environment value defined.

Reply:
I think there's a slight misunderstanding in the way this works. The only way to get config is via `kwargs` or environment variables so there is no difference (from a config priority point of view) between someone creating a routing rule that creates specific config and someone who doesn't.

Standard practice is to use environment variables as highest precedence which override all other config values. Sensible defaults should be set in the plugin and bespoke config values can be set in `alertad.conf` or the "routing rules".

I think that would work as most people would expect, and if this config precedence doesn't work for you then I would suggest that you aren't using environment variables correctly.

Happy to continue this discussion on https://gitter.im/alerta/chat.